## How to convert rna fusion supporting reads transcriptome alignmnent BAM to genome alignment BAM

1. get svid from the fusion report tsv file of the fusion gene
2. get the path of fusion supporting bam svscan generated for the sample
3. use ```svtools svbam``` to output the result you like

```
Usage: /share/work1/wulj/programs/bin/svtools svbam [OPTIONS]

Options:
  -h,--help                                       Print this help message and exit
  -i,--ibam TEXT:FILE REQUIRED                    bam generated by svscan
  -s,--svid INT REQUIRED                          SV ID
  -o,--obam TEXT=./svi.bam                        output bam path
  -O,--obam2 TEXT                                 output realign bam path
  -r,--ref TEXT                                   reference used to realign
```
You just need to set ```-r``` to BWA index path of genome, and ```-O``` to the genome BAM file you want to output to for the RNA fusion supporting reads.<br> <br>Open igv and do what ever you like, you may need use split view and ```Regions->Region Navigator``` to define multiple region to view fusion supporting status.
